### PR TITLE
allow non-minimal equation for padic heights

### LINKED ISSUE #34790: padic height calculations assume minimal equations
<div id="comment:0"></div>

[This mathoverflow question](https://mathoverflow.net/questions/433762/does-the-p-adic-regulator-depend-on-weierstrass-model) and [this asksage question](https://ask.sagemath.org/question/64632/p-adic-regulator-on-different-weierstrass-models/) raise an issue with p-adic heights. The algorithm currently assumes silently that the equation is minimal.

Here is the incorrect results pointed out in the questions. The regulator must be independent of the model used:

```
sage: E = EllipticCurve([-1728,-100656])
sage: E.padic_regulator(7)
3*7 + 3*7^2 + 4*7^3 + 3*7^4 + 5*7^6 + 7^8 + 3*7^9 + 5*7^11 + 7^12 + 2*7^13 + 3*7^15 + 3*7^16 + 4*7^17 + 6*7^18 + 5*7^19 + O(7^20)
sage: Emin = E.global_minimal_model()
sage: Emin.padic_regulator(7)
2*7^2 + 5*7^3 + 7^4 + 5*7^5 + 3*7^6 + 5*7^7 + 7^8 + 5*7^9 + 5*7^10 + 6*7^11 + 3*7^12 + 5*7^13 + 7^14 + 6*7^15 + 4*7^16 + 3*7^17 + 5*7^18 + 6*7^19 + O(7^20)
```

The problem is that the algorithm needs the points on which it evaluates the p-adic height to have good reduction in the model it works with.

One solution would be to raise a `NotImplementedError`. 

But better, especially if one wants to generalise this to number fields where elliptic curve do not have minimal models, one should find the correct multiple to use.

Component: **elliptic curves**

Author: **Chris Wuthrich**

Branch/Commit: **[u/wuthrich/trac34790](https://github.com/sagemath/sagetrac-mirror/tree/u/wuthrich/trac34790) @ [`a12a6d0`](https://github.com/sagemath/sagetrac-mirror/commit/a12a6d0657db7969e5b8fc744da69282aa8851c5)**

_Issue created by migration from https://trac.sagemath.org/ticket/34790_

### PR DESCRIPTION
**Pull request by chriswuthrich**
### :books: Description

Fixes #34790

The `p`-adic height calculations on an elliptic curve assumed that the equation was minimal.
This is fixed as well as some docstring formatting.

The main mathematical change is in padics.py, new line 579, which is then used in new line 931. This makes sure that the point is multiplied by an integer such that it has good reduction at all primes, even if the model is not minimal. Though it is still assumed that the models is integral.

The rest is cosmetics.

### :memo: Checklist

- `[x]` The title is concise, informative, and self-explanatory.
- `[x]` The description explains in detail what this PR is about.
- `[x]` I have linked a relevant issue or discussion.
- `[x]` I have created tests covering the changes.
- `[x]` I have updated the documentation accordingly.

(My first pull request, tell me if I am doing something wrong or where I need to read up on how we do things now.)

### CODE REVIEW

**Comment by chriswuthrich** [comment]
I don't understand the error in pycodestyle. Line 332 in constructor.py is ```When constructing a curve over `\QQ` from a Cremona or LMFDB``` and that has no "\Q" but a "\QQ" as it should.

**Comment by fchapoton** [comment]
> I don't understand the error in pycodestyle. Line 332 in constructor.py is `` When constructing a curve over `\QQ` from a Cremona or LMFDB `` and that has no "\Q" but a "\QQ" as it should.

You just need to start any doctest containing latex macros with `r"""` instead of `"""`.

**Comment by chriswuthrich** [comment]
> You just need to start any doctest containing latex macros with `r"""` instead of `"""`.

Oh, of course, thanks a lot for the help, Frédéric. I am a bit rusty.

**Comment by chriswuthrich** [comment]
This is ready for review in component "elliptic curves". (Can I add labels myself?)
Maybe @JohnCremona knows who could be ready to look at the changes.

**Comment by JohnCremona** [comment]
I will review this

File: src/sage/schemes/elliptic_curves/padics.py [hunk 1]
```diff
@@ -584,12 +576,68 @@ def _multiply_point(E, R, P, m):
...
+    else:
+        # generalising to number fields one can get the u from local_data
+        Emin = E.global_minimal_model()
+        iota = E.isomorphism_to(Emin)
+        u = Integer(iota.u)
+        ps = u.prime_divisors()
+        li = []
+        for p in ps:
+            np = u.valuation(p)
+            if Emin.discriminant() %p != 0:
```
Comment (JohnCremona) [inline src/sage/schemes/elliptic_curves/padics.py:643]: You could omit the !=0

File: src/sage/schemes/elliptic_curves/padics.py [hunk 2]
```diff
@@ -584,12 +576,68 @@ def _multiply_point(E, R, P, m):
...
+                li.append(Emin.Np(p) * p**(np-1))
+            elif Emin.has_additive_reduction(p):
+                li.append(E.tamagawa_number(p) * p**np)
+            elif E.has_split_multiplicative_reduction(p):
+                li.append(E.tamagawa_number(p) * (p-1) * p**(np-1))
+            else: # non split
+                li.append(E.tamagawa_number(p) * (p+1) * p**(np-1))
+        otherbad = Integer(Emin.discriminant()).prime_divisors()
+        otherbad = [p for p in otherbad if u%p != 0 ]
```
Comment (JohnCremona) [inline src/sage/schemes/elliptic_curves/padics.py:652]: You could omit the !=0

File: src/sage/schemes/elliptic_curves/padics.py [hunk 3]
```diff
@@ -584,12 +576,68 @@ def _multiply_point(E, R, P, m):
...
+        sage: _multiple_to_make_good_reduction(E)
+        30
+
+    The number ``n_2`` is not always optimal but it is in this example.
+    The first multiple of the generator `P` with good reduction in this
+    non-minimal model is `30 P`.
+
+    """
+    if not E.is_integral():
+        st = ("This only implemented for integral models. "
```
Comment (JohnCremona) [inline src/sage/schemes/elliptic_curves/padics.py:629]: add a doctest for this error

File: src/sage/schemes/elliptic_curves/padics.py [hunk 4]
```diff
@@ -584,12 +576,68 @@ def _multiply_point(E, R, P, m):
...
+        ps = u.prime_divisors()
+        li = []
+        for p in ps:
+            np = u.valuation(p)
+            if Emin.discriminant() %p != 0:
+                li.append(Emin.Np(p) * p**(np-1))
+            elif Emin.has_additive_reduction(p):
+                li.append(E.tamagawa_number(p) * p**np)
+            elif E.has_split_multiplicative_reduction(p):
+                li.append(E.tamagawa_number(p) * (p-1) * p**(np-1))
```
Comment (JohnCremona) [inline src/sage/schemes/elliptic_curves/padics.py:648]: perhaps add a doctest for this (split multiplicative) branch

File: src/sage/schemes/elliptic_curves/padics.py [hunk 5]
```diff
@@ -584,12 +576,68 @@ def _multiply_point(E, R, P, m):
...
+            else: # non split
+                li.append(E.tamagawa_number(p) * (p+1) * p**(np-1))
```
Comment (JohnCremona) [inline src/sage/schemes/elliptic_curves/padics.py:650]: again, perhaps add a doctest for this

**Review by JohnCremona — CHANGES_REQUESTED** [review]
Looks very good.  Th ecomments are all trivial, partly to keep codecov happy

File: src/sage/schemes/elliptic_curves/padics.py (hunk 1 above)
Comment (chriswuthrich) [reply src/sage/schemes/elliptic_curves/padics.py:643]: I prefer keeping it. Old habit to think like a mathematician: I think they increase the readability of the code, but don't slow it down significantly. But you insist, I am ok taking it off.

File: src/sage/schemes/elliptic_curves/padics.py (hunk 3 above)
Comment (chriswuthrich) [reply src/sage/schemes/elliptic_curves/padics.py:629]: imho: raise ValueError for arguments of the wrong type should not have a test. The worst that could happen is that we get the wrong kind of error.
But I will add one.

**Comment by chriswuthrich** [comment]
The added commits contain the requested doctests in padics.py. And it corrects my earlier changes of "√élu". It now  reads "square-root Vélu" instead. A discussion on this issue took place in the pull request #35626 . I am open for a larger discussion of this question and change it back to  "√élu".

**Comment by chriswuthrich** [comment]
I have clicked on "resolved" on the three reviewer's comments above that I think I have resolved. I don't know if that is my job or the reviewers' job.

**Comment by JohnCremona** [comment]
Probably my job to mark them as resolved, I will look again soon

File: src/sage/schemes/elliptic_curves/padics.py (hunk 1 above)
Comment (JohnCremona) [reply src/sage/schemes/elliptic_curves/padics.py:643]: I don't insist (I do the same) but some python purists might

File: src/sage/schemes/elliptic_curves/padics.py (hunk 2 above)
Comment (JohnCremona) [reply src/sage/schemes/elliptic_curves/padics.py:652]: as above

**Review by JohnCremona — APPROVED** [review]

**Comment by vbraun** [comment]
Merge conflict

**Comment by JohnCremona** [comment]
I'll switch this back to +ve review when tests pass

**Comment by vbraun** [comment]
Merge conflict

**Pull request merged**